From 171d90253f35b1826ec03bda48f1f02015660d47 Mon Sep 17 00:00:00 2001
From: Haibo Chen <haibo.chen@nxp.com>
Date: Fri, 20 Jul 2018 15:45:12 +0800
Subject: [PATCH 4239/5242] MLK-18721-1: input: touch: Synaptics: add x/y
 diagonal rotation support

commit  a2f2ac513e7f07925bcc7519e1816b1c95350099 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add x/y coordinate diagonal rotation support, user can use this feature
by add "synaptics,diagonal-rotation" in dts.

Signed-off-by: Haibo Chen <haibo.chen@nxp.com>
Reviewed-by: Fugang Duan <fugang.duan@nxp.com>
(cherry picked from commit 10c98865e5d1f0b7938523c793165e72add39f10)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../touchscreen/synaptics_dsx/synaptics_dsx_i2c.c  |   12 ++++++++++++
 .../touchscreen/synaptics_dsx/synaptics_dsx_i2c.h  |    1 +
 2 files changed, 13 insertions(+)

diff --git a/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.c b/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.c
index 415ea94..35fc226 100644
--- a/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.c
+++ b/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.c
@@ -1092,6 +1092,11 @@ static int synaptics_rmi4_f11_abs_report(struct synaptics_rmi4_data *rmi4_data,
 
 			x = (data.x_position_11_4 << 4) | data.x_position_3_0;
 			y = (data.y_position_11_4 << 4) | data.y_position_3_0;
+			if (rmi4_data->diagonal_rotation) {
+				x = rmi4_data->sensor_max_x - x;
+				y = rmi4_data->sensor_max_y - y;
+			}
+
 			wx = data.wx;
 			wy = data.wy;
 
@@ -1270,6 +1275,11 @@ static int synaptics_rmi4_f12_abs_report(struct synaptics_rmi4_data *rmi4_data,
 
 			x = (finger_data->x_msb << 8) | (finger_data->x_lsb);
 			y = (finger_data->y_msb << 8) | (finger_data->y_lsb);
+			if (rmi4_data->diagonal_rotation) {
+				x = rmi4_data->sensor_max_x - x;
+				y = rmi4_data->sensor_max_y - y;
+			}
+
 #ifdef REPORT_2D_W
 			wx = finger_data->wx;
 			wy = finger_data->wy;
@@ -3228,6 +3238,8 @@ static int synaptics_rmi4_probe(struct i2c_client *client,
 	rmi4_data->sleep_enable = synaptics_rmi4_sleep_enable;
 
 	rmi4_data->i2c_addr = client->addr;
+	rmi4_data->diagonal_rotation = of_property_read_bool(np,
+					"synaptics,diagonal-rotation");
 
 	mutex_init(&(rmi4_data->rmi4_io_ctrl_mutex));
 	mutex_init(&(rmi4_data->rmi4_reset_mutex));
diff --git a/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.h b/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.h
index d2d7381..12d7bf4 100644
--- a/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.h
+++ b/drivers/input/touchscreen/synaptics_dsx/synaptics_dsx_i2c.h
@@ -306,6 +306,7 @@ struct synaptics_rmi4_data {
 	int irq;
 	int sensor_max_x;
 	int sensor_max_y;
+	bool diagonal_rotation;
 	bool flash_prog_mode;
 	bool irq_enabled;
 	bool touch_stopped;
-- 
1.7.9.5

